VisualizationBase/src/items/ViewItem.cpp

```diff
@@ -102,38 +116,71 @@ const QPoint ViewItem::positionOfNode(Model::Node *node) const
 	return QPoint(-1, -1);
 }
 
-QVector<QVector<Model::Node*>> ViewItem::nodesGetter()
+const QPoint ViewItem::positionOfItem(Item *item) const
 {
-	return nodes_;
+	auto vref = DCast<VNodeReference>(item);
+	if (vref)
+		return positionOfNode(vref->node());
+	else return QPoint(-1, -1);
 }
 
-void ViewItem::ensureColumnExists(int column)
+Model::Node* ViewItem::nodeAt(int column, int row)
 {
-	if (nodes_.size() <= column)
-		nodes_.resize(column + 1);
+	if (column < 0 || column >= nodes_.size())
+		return nullptr;
+	if (row < 0 || row >= nodes_[column].size())
+		return nullptr;
+	return nodes_[column][row];
 }
 
-QString ViewItem::debugNodes()
+void ViewItem::addEmptyNode(int column, int row, int desiredYPosOfNextItem)
 {
-	//Prints the node grid as a debug string
-	QString result;
-	int maxHeight = 0;
-	for (int i = 0; i < nodes_.size(); i++)
-		if (nodes_.at(i).size() > maxHeight)
-			maxHeight = nodes_.at(i).size();
+	emptyNodesToBuild_.append(EmptyNodeParameters(column, row, desiredYPosOfNextItem));
+	setUpdateNeeded(StandardUpdate);
+}
 
-	for (int row = 0; row < maxHeight; row++)
+void ViewItem::updateGeometry(int availableWidth, int availableHeight)
+{
+	Super::updateGeometry(availableWidth, availableHeight);
+	while (!emptyNodesToBuild_.isEmpty())
 	{
-		for (int col = 0; col < nodes_.size(); col++)
+		auto node = emptyNodesToBuild_.takeFirst();
+		auto curHeight = 0;
+		if (node.row > 0)
 		{
-			if (nodes_.at(col).size() > row)
-				result += nodes_.at(col).at(row) ? (nodes_.at(col).at(row)->typeName() + " ") : "null ";
-			else
-				result += "null ";
+			auto previousItem = findVisualizationOf(nodeAt(node.column, node.row - 1));
+			curHeight = previousItem->scenePos().y() + previousItem->heightInParent();
 		}
-		result += "\n";
+		//TODO@cyril Would be nice not to hardcode the "-20", it takes the margin of the grid into account
+		auto nodeHeight = node.desiredYPosOfNextItem - 20 - curHeight;
+		auto empty = new Model::EmptyNode(nullptr);
+		empty->setCustomSize(0, nodeHeight);
+		ensureColumnExists(node.column);
+		auto ref = new Model::NodeReference();
+		ref->setTarget(empty);
```

Comment:
Setting the target to an emptyNode is a bad idea. But anyway we'll change this code.